Add new `.odc.explore()` method for automatically plotting raster data on an interactive map

robbibt opened the pull request:
This PR addresses #104 by adding a new `.explore()` method for automatically plotting raster data into an interactive map. This is designed to replicate the extremely useful `.explore()` method from Geopandas: https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.explore.html

Essentially `explore` combines `to_rgba` and `add_to` from this repo into a single easy to use method.

Supports single band `xr.DataArray` inputs:
![image](https://github.com/opendatacube/odc-geo/assets/17680388/28513cc7-7bfb-421b-9e28-9758bfc00d97)

Automatically converts `xr.Dataset` inputs to RGB:
![image](https://github.com/opendatacube/odc-geo/assets/17680388/73c62df6-9c11-49d2-ae76-c507b6eca5d0)

And easily passing in custom mapping params like basemaps:
![image](https://github.com/opendatacube/odc-geo/assets/17680388/7a49a962-0245-4b0e-bde4-a505ed67b566)

robbibt commented:
@Kirill888 I wasn't sure whether to define the main `explore` func in `_map.py` or `_xr_interop.py`. It might be better in `_xr_interop.py` so we can more easily include it in the docs, e.g. `odc.geo.xr.explore`?

robbibt committed:
Add new .odc.explore() method

Kirill888 commented on odc/geo/_map.py:
```diff
@@ -153,3 +153,141 @@ def add_to(
+    reproject_kwds.update(resampling=resampling)
+
+    # If input is a dataset, convert to an RGBA array
+    if isinstance(xx, xr.Dataset):
+        try:
+            xx = xx.odc.to_rgba(bands=bands, vmin=vmin, vmax=vmax)
+        except ValueError as e:
+            raise ValueError(
+                f"Unable to automatically guess RGB colours ({e}). "
+                f"Manually specify bands to plot using the `bands` "
+                "parameter, e.g. `ds.odc.explore(bands=['a', 'b', 'c'])`"
+            ) from e
```
`.add_to` does all of that and does it after all the projection and max size shrinkage. Code here should not perform any colorization and projection change, instead it should call `.add_to` with appropriate parameters.

Kirill888 replied:
In fact `.add_to(..., map=None)` will give you image data url and map bounds. So if you desire some customization of imagelayers and the like, you can just focus on that.

Code in explore should then do:

1. Check needed libs are installed
   - Ideally raise "actionable" error (Please install ...)
2. Convert Dataset/DataArray to an image+bounds using `.add_to`
    - All projection and colorization happens in there
    - Fill in some defaults that `.add_to` might not know about (picking the right kind of bands for example)
3. Create appropriate layers to add to a map and a map itself if not provided

robbibt replied:
Thanks @Kirill888! Looks like I can simplify this a fair bit - I didn't realise that `add_to` did its own reprojection. I did however reimplement some of the error catching (e.g. too many dims) outside of `.add_to` was that I was finding it would only raise errors _after_ reprojection and loading into memory - which would cause issues if users passed in e.g. hundreds of timesteps instead of one (data would load slowly then only return the error at the end). I'll have another look and see if I was doing something wrong to cause that to happen.

A few quick questions:
1. I would like to be able to support custom resampling when adding to the map, but [`.add_to` doesn't seem to support this](https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_map.py#L138). Any objection to me adding a `resampling` param to `.add_to`?
2. Does `.add_to` automatically convert `xr.Dataset` inputs to an RGBA array? I can't see `.add_to` as a method on `xr.Dataset`, only [`xr.DataArray`](https://github.com/opendatacube/odc-geo/blob/a9fbfc53625502d14dc7f124470ae0d54eb52f60/odc/geo/_xr_interop.py#L851) (I'd like `.explore()` to work on both: automatically converting and plotting `xr.Dataset` as an multiband RGB image, and  plotting `xr.DataArray`s as a single-band colorised image).
3. What is the advantage of running `.add_to(..., map=None)`, rather than the current approach of simply creating the map first and doing `.add_to(..., map=m)`?

Kirill888 replied:
Extra features like resampling config, or more robust input verification should go into `.add_to`.

robbibt replied:
I've worked out the problem with Dask inputs: if you accidently pass in an array with e.g. hundreds of timesteps, it gets loaded into memory inside `compress` here:
https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_compress.py#L95-L96

But that Dask load happens _before_ the excess time dimension is caught in `_compress_image` here:
https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_compress.py#L26-L33

That causes a long wait for data to load, with users only receiving the "Expect 2 or 3 dimensional array" error at the very end. Ideally, we'd be able to raise that error much earlier, avoiding the need to wait until the end of a potentially large Dask load.

Kirill888 replied:
> I've worked out the problem with Dask inputs: if you accidently pass in an array with e.g. hundreds of timesteps, it gets loaded into memory inside `compress` here: https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_compress.py#L95-L96

So ideally compress should catch that before calling "compute", something like `_verify_can_compress(xx: xr.DataArray)` that should contain logic similar to:

https://github.com/opendatacube/odc-geo/blob/a9fbfc53625502d14dc7f124470ae0d54eb52f60/odc/geo/_compress.py#L23-L33

which is basically: array `.ndim in (2,3)` after squeezing out single sized dimensions.

Kirill888 replied:
> 2\. Does `.add_to` automatically convert `xr.Dataset` inputs to an RGBA array? I can't see `.add_to` as a method on `xr.Dataset`, only [`xr.DataArray`](https://github.com/opendatacube/odc-geo/blob/a9fbfc53625502d14dc7f124470ae0d54eb52f60/odc/geo/_xr_interop.py#L851) (I'd like `.explore()` to work on both: automatically converting and plotting `xr.Dataset` as an multiband RGB image, and  plotting `xr.DataArray`s as a single-band colorised image).

I see, let's keep `.add_to` as is then, and perform `Dataset+bands->DataArray` conversion inside `.explore`.

robbibt replied:
Sounds good, will put something together 👍

Kirill888 commented on odc/geo/_map.py:
```diff
@@ -153,3 +153,141 @@ def add_to(
+    if map is None:
+        map = Map(**map_kwds)
+
+    # If necessary, reproject to map CRS
+    _crs = map_crs(map)
+    layer_name = xx.name  # copy out name as it is lost in reprojection
+    if _crs != xx.odc.crs:
+        xx = xx.odc.reproject(_crs, **reproject_kwds)
+
```
leave this logic to `.add_to`

robbibt replied:
Yep, this seems to work well.

robbibt committed:
Simplify .explore()

robbibt commented:
@Kirill888 Have made the following changes:

- Simplified `.explore()` by using existing reprojection functionality in `.add_to()`
- Added `_verify_can_compress` to catch unsuitable dims prior to Dask load in `compress`
- Removed error catching from `_compress_image` now those errors are caught upstream
- Added a new `resample` param to `add_to()` to control reprojection resampling method
- Added a more user friendly `ModuleNotFoundError` if Folium isn't installed
- Improved error messages for auto-band name guessing inside `to_rgba()`
- Minor error message wording and documentation improvements

Kirill888 commented on odc/geo/_compress.py:
```diff
@@ -19,18 +19,27 @@
         bands = im.reshape(nc, h, w)
-    else:
-        raise ValueError(f"Expect 2 or 3 dimensional array got: {im.ndim}")
 
```
unless code is on the "hot-path", i.e. gets executed millions of times in a tight loop, it's best to keep these kind of checks. Reasons why:

- communicating assumptions of the code to the reader
- so called "defensive programming" that makes code more resilient to future changes or mistakes in other parts of the existing code. Just because right now this is only reachable through another function that performs that check, this won't always stay that way.

Maybe replacing `raise ValueError` with `assert` to indicate that it's the responsibility of the caller to perform such a check, but it's still important to have something that states data shape expectations even within "internal" functions like this one.

robbibt replied:
Thanks for the explanation, makes sense! Will put this back in

robbibt replied:
Done - just left it as a ValueError for now as it felt cleaner that way.

Kirill888 commented:
@robbibt thanks, this is great. Let's put back those data shape checks into `._compress` and it's good to merge after that.

robbibt committed:
Restore error handling in _compress_image

robbibt commented:
> @robbibt thanks, this is great. Let's put back those data shape checks into `._compress` and it's good to merge after that.

Awesome, thanks for the great feedback. Have made that change, can rebase merge once you approve.

What's the plan for the next `odc-geo` version release? I'd like to get `mask`, `crop` and `explore` into the DEA Sandbox and it would be much easier with a proper release.

Kirill888 reviewed (approved):
thanks, good to merge
